code/search/SearchIndex.php

```diff
@@ -109,8 +153,10 @@ public function fieldData($field, $forceType = null, $extraOptions = array())
                             if (!isset($options['origin'])) {
                                 $options['origin'] = $dataclass;
                             }
-                            $next[$class] = $options;
-                            continue 2;
+
+                            // we add suffix here to prevent the relation to be overwritten by other instances
+                            // all sources lookups must clean the source name before reading it via getSourceName()
+                            $next[$class . self::$class_delimiter . $dataclass] = $options;
```

Comment:
You've declared `$class_delimiter` as a configurable property, so might be worth using `self::config()->get('class_delimiter')` here instead of accessing the property directly